addons/mail/models/mail_thread.py

```diff
@@ -1821,36 +1824,38 @@ def message_post(self, *,
-                to the related document.
+            Should not be a list of commands. Attachment records attached to mail
+            composer will be attached to the related document.
 
         Extra keyword arguments will be used either
           * as default column values for the new mail.message record if they match
             mail.message fields;
-          * propagated to notification methods;
+          * propagated to notification methods if not;
 
         :return record: newly create mail.message
         """
         self.ensure_one()  # should always be posted on a record, use message_notify if no record
-        # split message additional values from notify additional values
-        msg_kwargs = dict((key, val) for key, val in kwargs.items() if key in self.env['mail.message']._fields)
-        notif_kwargs = dict((key, val) for key, val in kwargs.items() if key not in msg_kwargs)
 
         # preliminary value safety check
-        partner_ids = set(partner_ids or [])
-        if self._name == 'mail.thread' or not self.id or message_type == 'user_notification':
-            raise ValueError(_('Posting a message should be done on a business document. Use message_notify to send a notification to an user.'))
-        if 'channel_ids' in kwargs:
-            raise ValueError(_("Posting a message with channels as listeners is not supported since Odoo 14.3+. Please update code accordingly."))
-        if 'model' in msg_kwargs or 'res_id' in msg_kwargs:
-            raise ValueError(_("message_post does not support model and res_id parameters anymore. Please call message_post on record."))
-        if 'subtype' in kwargs:
-            raise ValueError(_("message_post does not support subtype parameter anymore. Please give a valid subtype_id or subtype_xmlid value instead."))
+        self._raise_for_invalid_fields([kwargs.keys()], {'model', 'res_id', 'subtype'})
+        if self._name == 'mail.thread' or not self.id:
+            raise ValueError(_("Posting a message should be done on a business document. Use message_notify to send a notification to an user."))
+        if message_type == 'user_notification':
+            raise ValueError(_("Use message_notify to send a notification to an user."))
+        partner_ids = list(dict.fromkeys(partner_ids or []))
+        attachment_ids = list(dict.fromkeys(attachment_ids or []))
         if any(not isinstance(pc_id, int) for pc_id in partner_ids):
             raise ValueError(_('message_post partner_ids and must be integer list, not commands.'))
+        if any(not isinstance(attach_id, int) for attach_id in attachment_ids):
+            raise ValueError(_('message_post attachment_ids and must be integer list, not commands.'))
+
+        # split message additional values from notify additional values
+        msg_kwargs = dict((key, val) for key, val in kwargs.items()
+                          if key in self.env['mail.message']._fields)
```

Comment:
Don't we usually prefer comprehensions? (Oh, like in L2273)
```suggestion
        msg_kwargs = {key: val for key, val in kwargs.items()
                      if key in self.env['mail.message']._fields)
```

Reply:
"We", maybe. "I" prefer dict :D . But ok, I may accept to evolve.